### PR TITLE
Start and Shutdown VM Notification

### LINKED ISSUE #695: toast notifications when VM is created, starting, or shutting down
As a VM Portal user, I want to see toast notifications when a VM is starting or shutting down, like in the Admin Portal, so that I can be certain the action is occurring [feedback].
image

### PR DESCRIPTION
**Pull request by yodem**
fix #695 

Request to Start VM notification -
toastNotification1

Request to Shutdown VM notification -
toastNotification2

The way you toast a massage is now generic, so far in src/components/ToastNotifications.js,
the notification type was only warning, however now you can pass a value into it.

### CODE REVIEW

**Comment by yodem** [comment]
Later on we need to check when these processes have finished and to toast another notification, and to add a notification when VM has created, but I understood that first of all we will wait for the wizard fiture to be ready and then we will implement it.

**Comment by yodem** [comment]
@sjd78 @bond95  What do you say?

**Comment by yodem** [comment]
No problem i agree, @sjd78 @michalskrivanek @lwrigh do you think its a good idea?

**Comment by yodem** [comment]
> I would suggest making the toast notification text more conversational. Here are my suggestions for making edits to the toast notification text:
> Start: "Request to start VM has been received."
> Shutdown: "Request to shutdown VM has been received."
> Restart: "Request to restart VM has been received."
> Suspend: "Request to suspend VM has been received."
> Allocating VM from a new pool: "Request to allocate VM from a new pool has been received."
> 
> I also question whether we need to present the user with any more toast notifications? Do you think it's enough that the VM status changes and the user notices that change? I think this is a scenario where we don't need to present a toast notification to the user since it's such a short lived action.

I think we need to add a notification once the vm is finally up\down.

**Comment by lwrigh** [comment]
@michalskrivanek Does admin portal present the user with a toast notification after it is first up and running?

File: src/components/ToastNotifications.js [hunk 1]
```diff
@@ -9,8 +9,9 @@ import style from './sharedStyle.css'
 
 const ToastNotifications = ({ userMessages, onDismissNotification }) => {
   return <ToastNotificationList>
+
     { userMessages.get('records').filter(r => !r.get('notified')).map(r =>
-      <TimedToastNotification className={style['toast-margin-top']} type='warning' onDismiss={() => onDismissNotification(r.get('time'))} key={r.get('time')}>
+      <TimedToastNotification className={style['toast-margin-top']} type={r.get('type') ? `${r.get('type')}` : 'warning'} onDismiss={() => onDismissNotification(r.get('time'))} key={r.get('time')}>
```
Comment (sjd78) [inline src/components/ToastNotifications.js]:
```suggestion
      <TimedToastNotification className={style['toast-margin-top']} type={r.get('type') || 'warning'} onDismiss={() => onDismissNotification(r.get('time'))} key={r.get('time')}>
```

or....
```suggestion
      <TimedToastNotification className={style['toast-margin-top']} type={r.get('type') ? r.get('type') : 'warning'} onDismiss={() => onDismissNotification(r.get('time'))} key={r.get('time')}>
```

File: src/sagas/index.js [hunk 2]
```diff
@@ -689,6 +690,9 @@ export function* stopProgress ({ vmId, poolId, name, result }) {
 function* shutdownVm (action) {
   yield startProgress({ vmId: action.payload.vmId, name: 'shutdown' })
   const result = yield callExternalAction('shutdown', Api.shutdown, action)
+  if (result.status === 'complete') {
+    yield put(addUserMessage({ message: 'VM Shutdown Request Received', type: 'success' }))
```
Comment (sjd78) [inline src/sagas/index.js]: The text needed to be externalized (moved to `messages.js`).

I suggest naming the message key `toastVmShutdown`.

File: src/sagas/index.js [hunk 3]
```diff
@@ -708,6 +712,9 @@ function* startVm (action) {
   yield startProgress({ vmId: action.payload.vmId, name: 'start' })
   const result = yield callExternalAction('start', Api.start, action)
   // TODO: check status at refresh --> conditional refresh wait_for_launch
+  if (result.status === 'complete') {
+    yield put(addUserMessage({ message: 'VM Start Request Received', type: 'success' }))
```
Comment (sjd78) [inline src/sagas/index.js]: The text needed to be externalized (moved to `messages.js`).

I suggest naming the message key `toastVmStart`.

**Comment by sjd78** [comment]
@yodem - All of the message strings need to be externalized (i.e. moved to `messages.js`).

File: src/sagas/index.js [hunk 4]
```diff
@@ -657,31 +660,46 @@ export function* stopProgress ({ vmId, poolId, name, result }) {
 function* shutdownVm (action) {
   yield startProgress({ vmId: action.payload.vmId, name: 'shutdown' })
   const result = yield callExternalAction('shutdown', Api.shutdown, action)
+  if (result.status === 'complete') {
+    yield put(addUserMessage({ message: msg.actionFeedbackShutdownVm({ VmId: action.payload.vmId }), type: 'success' }))
```
Comment (sjd78) [inline src/sagas/index.js]: Looking good, almost there.  Using the VM Name would be more appropriate here.  The only place I remember seeing the VM id itself is in the URL for the details page.

In a saga, given the VM id, you can get the VM name like this:
```jsx
const vmName = yield select(state => state.vms.getIn([ 'vms', vmId, 'name' ]))
```

File: src/sagas/index.js [hunk 5]
```diff
@@ -657,31 +660,46 @@ export function* stopProgress ({ vmId, poolId, name, result }) {
...
+    yield put(addUserMessage({ message: msg.actionFeedbackStartVm({ VmId: action.payload.vmId }), type: 'success' }))
+  }
   yield stopProgress({ vmId: action.payload.vmId, name: 'start', result })
 }
 
 function* startPool (action) {
   yield startProgress({ poolId: action.payload.poolId, name: 'start' })
   const result = yield callExternalAction('startPool', Api.startPool, action)
+  if (result.status === 'complete') {
+    yield put(addUserMessage({ message: msg.actionFeedbackAllocateVm({ poolname: action.payload.poolId }), type: 'success' }))
```
Comment (sjd78) [inline src/sagas/index.js]: Pool name is a little different than VM name...

```jsx
const poolName = yield select(state => state.vms.getIn([ 'pools', poolId, 'name' ]))
```

File: src/components/ToastNotifications.js [hunk 6]
```diff
@@ -7,10 +7,12 @@ import { setNotificationNotified } from '_/actions'
 
 import style from './sharedStyle.css'
 
+const notificationTypes = ['error', 'warning', 'success', 'info', 'danger']
 const ToastNotifications = ({ userMessages, onDismissNotification }) => {
   return <ToastNotificationList>
+
     { userMessages.get('records').filter(r => !r.get('notified')).map(r =>
-      <TimedToastNotification className={style['toast-margin-top']} type='warning' onDismiss={() => onDismissNotification(r.get('time'))} key={r.get('time')}>
+      <TimedToastNotification className={style['toast-margin-top']} type={notificationTypes.includes(r.get('type')) ? `${r.get('type')}` : 'warning'} onDismiss={() => onDismissNotification(r.get('time'))} key={r.get('time')}>
```
Comment (sjd78) [inline src/components/ToastNotifications.js]: - This is a case sensitive check ... maybe `notificationTypes.includes(r.get('type').toLowerCase()`
 - The code `` `${r.get('type')}` `` can be reduced to just `r.get('type')`.

**Comment by sgratch** [comment]
@yodem can you please add a screenshot of how the notifications look now? Is https://github.com/oVirt/ovirt-web-ui/pull/1110#issue-313217711 up to date?

**Comment by sgratch** [comment]
@yodem There is a conflict on src/components/ToastNotifications.js, please resolve it

**Comment by yodem** [comment]
@sgratch Done.

**Comment by sjd78** [comment]
rebase LGTM

**Comment by isaranova** [comment]
LGTM

**Pull request merged**